File: solutions/ruby/acronym/5/acronym.rb
module Acronym

  class << self

    private

    STEPS = [
      ->(text) { text.tr("-", ' ') },
      ->(text) { text.upcase },
      ->(text) { text.gsub(/[[:punct:]]/, '') },
      ->(text) { text.split },
      ->(words) { words.map(&:chr) }
    ]

    public

    def abbreviate(expansion)
      STEPS.reduce(expansion) { |acc, step| step.call(acc) }.join
    end

  end

end
      

File: solutions/elixir/acronym/5/lib/acronym.ex
defmodule Acronym do

  defp words(phrase) do
    String.split(phrase,[" ", "_", "-"])
  end

  def abbreviate(phrase) when is_binary(phrase) do
    phrase
    |> words()
    |> Enum.map_join(&String.first/1)
    |> String.upcase()
  end

end
